storage/bootstrap/bootstrapper/base.go

```diff
@@ -34,77 +35,118 @@ const (
 // baseBootstrapper provides a skeleton for the interface methods.
 type baseBootstrapper struct {
 	opts bootstrap.Options
-	s    bootstrap.Source
+	src  bootstrap.Source
 	next bootstrap.Bootstrapper
 }
 
 // NewBaseBootstrapper creates a new base bootstrapper.
 func NewBaseBootstrapper(
-	s bootstrap.Source,
+	src bootstrap.Source,
 	opts bootstrap.Options,
 	next bootstrap.Bootstrapper,
 ) bootstrap.Bootstrapper {
 	bs := next
 	if next == nil {
 		bs = NewNoOpNoneBootstrapper()
 	}
-	return &baseBootstrapper{opts: opts, s: s, next: bs}
+	return &baseBootstrapper{opts: opts, src: src, next: bs}
 }
 
-// Bootstrap performs bootstrapping for the given shards and the associated time ranges.
-func (bsb *baseBootstrapper) Bootstrap(shard uint32, targetRanges xtime.Ranges) (bootstrap.ShardResult, xtime.Ranges) {
-	if xtime.IsEmpty(targetRanges) {
+func (b *baseBootstrapper) Can(strategy bootstrap.Strategy) bool {
+	return b.src.Can(strategy)
+}
+
+func (b *baseBootstrapper) Bootstrap(shardsTimeRanges bootstrap.ShardTimeRanges) (bootstrap.Result, error) {
+	if shardsTimeRanges.IsEmpty() {
 		return nil, nil
 	}
 
-	availableRanges := bsb.s.GetAvailability(shard, targetRanges)
-	remainingRanges := targetRanges.RemoveRanges(availableRanges)
+	available := b.src.Available(shardsTimeRanges)
+
+	remaining := make(map[uint32]xtime.Ranges)
+	for shard, ranges := range shardsTimeRanges {
```

Comment:
Can we move this logic to a separate method (e.g., `Subtract`) on the `ShardTimeRanges` interface?